File: 0245. Shortest Word Distance III/main0245.py
from typing import List


class Solution0245:
    def shortestWordDistance(self, wordsDict: List[str], word1: str, word2: str) -> int:
        wordLen = len(wordsDict)
        prev = -1
        isEqual = (word1 == word2)
        res = wordLen
        for i in range(wordLen):
            if wordsDict[i] == word1 or wordsDict[i] == word2:
                if prev != -1 and (isEqual or wordsDict[prev] != wordsDict[i]):
                    res = min(res, i - prev)
                prev = i
        
        return res

wordsDict = ["practice", "makes", "perfect", "coding", "makes"]
word1 = "makes"
word2 = "coding"
sol = Solution0245()
res = sol.shortestWordDistance(wordsDict, word1, word2)
print(res)

wordsDict = ["practice", "makes", "perfect", "coding", "makes"]
word1 = "makes"
word2 = "makes"
sol = Solution0245()
res = sol.shortestWordDistance(wordsDict, word1, word2)
print(res)

File: 0245. Shortest Word Distance III/main0245.java
public class main0245 {
    public static void main(String[] args) {
        Solution0245 sol = new Solution0245();
        String wordsDict[], word1, word2;
        int dist;

        wordsDict = new String[]{"practice", "makes", "perfect", "coding", "makes"};
        word1 = "makes";
        word2 = "coding";
        dist = sol.shortestWordDistance(wordsDict, word1, word2);
        System.out.println(dist);

        wordsDict = new String[]{"practice", "makes", "perfect", "coding", "makes"};
        word1 = "makes";
        word2 = "makes";
        dist = sol.shortestWordDistance(wordsDict, word1, word2);
        System.out.println(dist);        
    }
}


class Solution0245 {
    public int shortestWordDistance(String[] wordsDict, String word1, String word2) {
        int res = wordsDict.length;
        int prev = -1;
        for (int i = 0 ; i < wordsDict.length; i++) {
            if (wordsDict[i].equals(word1) || wordsDict[i].equals(word2)) {
                if (prev != -1 && (word1.equals(word2) || !wordsDict[prev].equals(wordsDict[i]))) {
                    res = Math.min(i - prev, res);
                }
                prev = i;
            }
        }
        return res;
    }
}